src/memory_constant.jl

```diff
@@ -0,0 +1,86 @@
+export CuConstantMemory
+
+# Map a constant memory name to its array value
+const constant_memory_initializer = Dict{Symbol,WeakRef}()
+
+"""
+    CuConstantMemory{T,N}(value::Array{T,N})
+    CuConstantMemory{T}(::UndefInitializer, dims::Integer...)
+    CuConstantMemory{T}(::UndefInitializer, dims::Dims{N})
+
+Construct an `N`-dimensional constant memory array of type `T`, where `isbits(T)`.
+
+Note that `deepcopy` will be called on the `value` constructor argument, meaning that
+mutations to the original `value` or its elements after construction will not be
+reflected in the value of `CuConstantMemory`.
+
+The `UndefInitializer` constructors behave exactly like the regular `Array` version,
+i.e. the value of `CuConstantMemory` will be completely random when using them.
+
+Unlike in CUDA C, structs cannot be put directly into constant memory. This feature can
+be emulated however by wrapping the struct inside of a 1-element array.
+
+When using `CuConstantMemory` as a global variable it is required to pass it as an argument
+to a kernel, where the argument is of type [`CuDeviceConstantMemory{T,N}`](@ref).
+When using `CuConstantMemory` as a local variable that is captured by a kernel closure
+this is not required, and it can be used directly like any other captured variable
+without passing it as an argument.
+
+In cases where the same kernel object gets called mutiple times, and it is desired to mutate
+the value of a `CuConstantMemory` variable in this kernel between calls, please refer
+to [`Base.copyto!(const_mem::CuConstantMemory{T}, value::Array{T}, kernel::HostKernel)`](@ref)
+"""
+struct CuConstantMemory{T,N} <: AbstractArray{T,N}
+    name::Symbol
+    size::Dims{N}
+    value::Union{Nothing,Array{T,N}}
+
+    function CuConstantMemory(value::Array{T,N}) where {T,N}
+        Base.isbitstype(T) || throw(ArgumentError("CuConstantMemory only supports bits types"))
+        name = gensym("constant_memory")
```

Comment:
A problem with this is that every invocation of a function that launches a kernel with constant memory, will result in a recompilation (because the name is tied to the instance).